### PR TITLE
KEYCLOAK-10169 OpenShift 4 Identity Provider

### PR DESCRIPTION
**Pull request by slaskawi**
https://issues.jboss.org/browse/KEYCLOAK-10169

### CODE REVIEW

**Comment by slaskawi** [comment]
This needs more testing. Closing for now.

**Comment by slaskawi** [comment]
Reopening and uploading new patch.

**Comment by slaskawi** [comment]
Hey @davidfestal 

Here's a patch for the Keycloak codebase that introduces support for OpenShift v4 Identity Providers. Please have a look at it when you find a moment.

**Comment by slaskawi** [comment]
@keycloak-ci-bot test

**Review by davidfestal — CHANGES_REQUESTED** [review]
Hi @slaskawi. Sorry for the delay. Code seems good to me.
Is there any Keycloak docker image I can use with this code included, in order to test it on our context, and validate our whole scenario works as it used to work when using the standalone openshift-v4 repo ?

**Comment by matskiv** [comment]
+1 on docker image, I would like to test it in Integreatly context as soon as possible :)

**Comment by slaskawi** [comment]
@matskiv @davidfestal No problem, here's the docker image to try it out `docker run -e KEYCLOAK_USER=admin -e KEYCLOAK_PASSWORD=admin slaskawi/keycloak-openshift-4-full`.

**Review by davidfestal — APPROVED** [review]
I tested with he provided docker image, and everything seems to work fine, at least in the Che use case.

**Review by davidfestal — CHANGES_REQUESTED** [review]
Well, that's not fully true: I have this error when trying to start a workspace:

```
[0m[33m15:47:18,115 WARN  [org.keycloak.events] (default task-12) type=IDENTITY_PROVIDER_RETRIEVE_TOKEN_ERROR, realmId=17112a72-622c-42ac-a77b-261f69b54781, clientId=null, userId=null, ipAddress=18.130.160.55, error=Client [che-public] not authorized to retrieve tokens from identity provider [openshift-v4].
[0m[31m15:47:18,115 ERROR [org.keycloak.services.resources.IdentityBrokerService] (default task-12) Client [che-public] not authorized to retrieve tokens from identity provider [openshift-v4].
[0m[33m15:49:45,394 WARN  [org.keycloak.events] (default task-11) type=IDENTITY_PROVIDER_RETRIEVE_TOKEN_ERROR, realmId=17112a72-622c-42ac-a77b-261f69b54781, clientId=null, userId=null, ipAddress=18.130.160.55, error=Client [che-public] not authorized to retrieve tokens from identity provider [openshift-v4].
[0m[31m15:49:45,395 ERROR [org.keycloak.services.resources.IdentityBrokerService] (default task-11) Client [che-public] not authorized to retrieve tokens from identity provider [openshift-v4].
[0m[33m15:50:00,208 WARN  [org.keycloak.events] (default task-11) type=IDENTITY_PROVIDER_RETRIEVE_TOKEN_ERROR, realmId=17112a72-622c-42ac-a77b-261f69b54781, clientId=null, userId=null, ipAddress=18.130.160.55, error=Client [che-public] not authorized to retrieve tokens from identity provider [openshift-v4].
[0m[31m15:50:00,208 ERROR [org.keycloak.services.resources.IdentityBrokerService] (default task-11) Client [che-public] not authorized to retrieve tokens from identity provider [openshift-v4].
```

I'll have to investigate more.

**Review by davidfestal — DISMISSED** [review]
Finally it was a false alert: When creating the federated identity from the Account page (with account linking only option), the read-token permission is not added to  the user by default for the broker client.
This is a very specific custom scenario that I was using, and is not the normal way to use Che with Keycloak. Additionally, this can be fixed very easily by manually adding the read-token permission to the user and relogin.

So everything seems to work as expected with Che.
+1 for me.

**Comment by nickboldt** [comment]
How can we rebuild SSO to include this patch? I've tried injecting the provider as a jar into the sso server [1], but I'd prefer to build it in directly via this PR. 

[1] https://gist.github.com/nickboldt/1b0002dd006aea827ece386bb4901447#3-create-dockerfile--build-image

**Comment by slaskawi** [comment]
@nickboldt In that case, it's probably better to use https://github.com/slaskawi/KEYCLOAK-10169-OpenShift4-User-Provider repo. @davidfestal could probably give you a hind with this. The procedure is the same as for the community (but you will probably need to put some Freemarker templates into the distribution).

**Comment by nickboldt** [comment]
Yes, that's what I was doing in https://gist.github.com/nickboldt/1b0002dd006aea827ece386bb4901447#1-build-openshift4-extension

But I was looking for a way to have THIS PR included too, rather than the external jar.

**Comment by slaskawi** [comment]
> But I was looking for a way to have THIS PR included too, rather than the external jar.

@nickboldt No, sorry. I don't think it's possible.

> Can you define "some Freemarker templates" ? Where? what? how? who?

I'm not exactly sure why, but I remember that someone had to put [this resources](https://github.com/slaskawi/KEYCLOAK-10169-OpenShift4-User-Provider/tree/master/src/main/resources/theme-resources/resources) into `$RHSSO/themes/base/admin/resources/partials`. I'm not sure what version it was and I'm not 100% you will need it. But just in case...

**Comment by nickboldt** [comment]
Yeah, we're already doing that...

```
git clone https://github.com/che-incubator/KEYCLOAK-10169-OpenShift4-User-Provider && cd KEYCLOAK-10169-OpenShift4-User-Provider
mvn clean install

# build SSO image with updated OS4 Provider
echo "
FROM registry.access.redhat.com/redhat-sso-7/sso73-openshift:1.0-12
COPY target/openshift4-extension-6.0.1.jar /opt/eap/standalone/deployments/
RUN touch /opt/eap/standalone/deployments/openshift4-extension-6.0.1.jar.dodeploy && \
    unzip -j /opt/eap/standalone/deployments/openshift4-extension-6.0.1.jar -d \
      /opt/eap/themes/base/admin/resources/partials \
      theme-resources/resources/realm-identity-provider-openshift-v4.html \
      theme-resources/resources/realm-identity-provider-openshift-v4-ext.html
" > Dockerfile
docker build . -t ${SSO_PATCH_IMAGE}
docker push ${SSO_PATCH_IMAGE}
```

**Comment by stianst** [comment]
This is on hold and we can not merge to master until we are able to automatically test it

**Comment by vmuzikar** [comment]
@slaskawi Is there a corresponding PR to the [keycloak-documentation](https://github.com/keycloak/keycloak-documentation) repo? (Couldn't find it.)

**Comment by slaskawi** [comment]
@vmuzikar At the moment there is none. I was planning to write something once this one goes in (just in case the implementation diverges during discussions on the pull request).

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/SocialLoginTest.java
Comment (vmuzikar) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/SocialLoginTest.java]: We could maybe add `testTokenExchange();`. The IdP supports it and test passes.

**Review by vmuzikar — COMMENTED** [review]
Overall, looks good to me, test passes – thanks @slaskawi!

Just 3 minor comments:
- It's really hard to configure the IdP properly. Error messages like the following doesn't tell much:
```
14:15:25,580 WARN  [org.keycloak.events] (default task-3) type=IDENTITY_PROVIDER_LOGIN_ERROR, realmId=8880964c-c014-4c62-92ac-b30fc079ba14, clientId=account, userId=null, ipAddress=127.0.0.1, error=couldNotSendAuthenticationRequestMessage, identity_provider=openshift-v4, code_id=0332dd5b-a839-4997-a043-4da5a708a902
14:15:25,581 ERROR [org.keycloak.services.resources.IdentityBrokerService] (default task-3) couldNotSendAuthenticationRequestMessage
```
I wouldn't be able to get it working without directly debugging the sources to see the exact Exceptions why it failed.
- The test automation is not ready yet – we need to prepare our CI environment for this first (OCP4 cluster and social-config.properties file). Until then `SocialLoginTest.openshift4Login` will need to be skipped.
- We must not forget to add some docs regarding this IdP.

**Comment by slaskawi** [comment]
> It's really hard to configure the IdP properly. Error messages like the following doesn't tell much:

@vmuzikar I bet you hit the SSL issue (OpenShift API URL is protected with TLS and we need a proper truststore). This Pull Request should address this: https://github.com/jboss-dockerfiles/keycloak/pull/201

> The test automation is not ready yet – we need to prepare our CI environment for this first (OCP4 cluster and social-config.properties file). Until then SocialLoginTest.openshift4Login will need to be skipped.

@vmuzikar I guess it's not a problem at least for now. Let's wait till we have a proper environment.

> We must not forget to add some docs regarding this IdP.

@vmuzikar I'm compiling a documentation paragraph as we speak. However, I'd like to make it up to date so I need to test everything again with the latest changes. Please stay tuned.

**Comment by vmuzikar** [comment]
> @vmuzikar I bet you hit the SSL issue (OpenShift API URL is protected with TLS and we need a proper truststore). This Pull Request should address this: [jboss-dockerfiles/keycloak#201](https://github.com/jboss-dockerfiles/keycloak/pull/201)

Yeah... among other issues (bad endpoint URL, etc.). :) That's why it was so hard to debug.

File: themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-openshift-v4.html
Comment (matskiv) [inline themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-openshift-v4.html]: btw: shouldn't this label have `for="storeToken"` ? :)

**Comment by slaskawi** [comment]
The documentation is provided here: https://github.com/keycloak/keycloak-documentation/pull/720

I also created a small demo, which you may watch here: https://www.youtube.com/watch?v=SZPDLuOr3ig. The sources for the demo are located here: https://github.com/slaskawi/keycloak-openshift4-demo

File: themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-openshift-v4.html
Comment (slaskawi) [reply themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-openshift-v4.html]: Of course! Thanks for spotting this!

**Comment by hmlnarik** [comment]
@vmuzikar Is this PR now ready from your point of view?

**Comment by vmuzikar** [comment]
@hmlnarik Yes, it's good to go but only if we merge the PR for skipping the test in the pipeline at the same time.

**Review by hmlnarik — APPROVED** [review]
Approving based on @vmuzikar 's review

**Comment by nickboldt** [comment]
So... will this fix be included in SSO 7.3.4 ? Just checking if that is still the plan. :)

**Comment by hmlnarik** [comment]
@nickboldt We cannot publicly comment on particular feature / bug fix availability in the product. Kindly check https://issues.jboss.org/browse/KEYCLOAK-10972 for your request.

**Pull request merged**
